README.md
# hopscotch

hopscotch is a tiny, high-performance HTTP image proxy. The idea is that you
put it behind your SSL terminator. When you serve user content you rewrite
image URLs to point to it. Just like that, all the assets in the page are
served over SSL and you avoid mixed content warnings.

It's inspired by [Camo](https://github.com/atmos/camo), Github's implementation
of the same concept.

## features

- returns 404 for all errors
- cache friendly
- transparently follow redirects
- limit content by size or MIME type
- paranoid mode that fails requests to internal networks

## setup

### docker

```bash
$ docker run --name hopscotch.psgi -e HOPSCOTCH_KEY=abc123 -d -p 8080:8080 bin/hopscotch.psgi
```

### carton

```bash
$ git clone http://github.com/fastmail/hopscotch.git
$ cd hopscotch
$ curl -L http://cpanmin.us | perl - Carton Starlet
$ carton install --deployment
$ env HOPSCOTCH_KEY=abc123 carton exec plackup -s Starlet -p 8080 ./bin/hopscotch.psgi
```

### anything else

Get Perl, get all the dependencies, run the program. This is the developer
option, and you're expected to know what you're doing :)

## URL format

hopscotch URLs look like:

    http://hopscotch/<digest>/<hexurl>

`<digest>` is the SHA256 HMAC digest generated from the unescaped image URL and
a secret key that you provide. `hexurl` is a simple hex encoding (two hex
digits per string byte, lowercase) of the same image URL used to generate the
digest.

Sample Perl code to generate a URL:

```perl
use Crypt::Mac::HMAC qw(hmac_hex);
my $hsurl = "http://hopscotch:8080/" . hmac_hex("SHA256", $KEY, $URL) . "/" . unpack("h*", $URL);
```

## config

hopscotch is configured with environment variables.

* `HOPSCOTCH_KEY`: shared key used to generate the HMAC digest (required)
* `HOPSCOTCH_HOST`: name to insert in the `x-hopscotch-host` header returned to
                    clients (default: `unknown`)
* `HOPSCOTCH_TIMEOUT`: max seconds hopscotch will wait for a response befores
                       giving up (default: 10)
* `HOPSCOTCH_HEADER_VIA`: string to include in the `via` header set to the
                          origin host (default: `hopscotch`)
* `HOPSCOTCH_HEADER_UA`: string to include in the `user-agent` header set to the
                         origin host (default: `hopscotch`)
* `HOPSCOTCH_LENGTH_LIMIT`: maximum `content-length` hopscotch will proxy (default:
                            5242880)
* `HOPSCOTCH_MAX_REDIRECTS`: maximum number of redirect hopscotch will follow
                             (default: 4)
* `HOPSCOTCH_PARANOID`: if set to a true value, hopscotch will refuse to connect
                        to "internal" networks as determined by
                        [`Net::DNS::Paranoid`](https://metacpan.org/pod/Net::DNS::Paranoid).
                        This costs an extra DNS lookup for each request but
                        protects against a class of SSRF attacks (default: 0)
* `HOPSCOTCH_ERRORS`: if set to a true value, reasons for proxy failures will
                      be output to stderr (default: true)
* `HOPSCOTCH_CAFILE`: file containing CA certificates for HTTPS verification (default: unset)
* `HOPSCOTCH_CAPATH`: dir containing CA certificates for HTTPS verification (default: unset)

If neither the `CAFILE` nor the `CAPATH`, the certificates provided by
`Mozilla::CA` will be used.

## who's using this?

* [FastMail](https://www.fastmail.fm/) are using this to serve images embedded in email

## credits and license

Copyright (c) 2014 Robert Norris. MIT license. See [LICENSE.md](LICENSE.md)

## contributing

Please hack on this and send pull requests :)



Changes
Revision history for Hopscotch

{{$NEXT}}
        - Hopscotch!


t/lib/Hopscotch/Tester.pm
use 5.010;
use strict;
use warnings;
package Hopscotch::Tester;

use LWP::Protocol::PSGI;
use LWP::UserAgent;
use MIME::Base64 qw(decode_base64);

my $baseport = 443;

my $host = 'hopscotch.local';
my $port = sub { $baseport++ };

{
    package Hopscotch::Tester::LWP::Wrapper;

    sub new {
        my ($class, $hostport, $guard) = @_;

        return bless {
            lwp          => LWP::UserAgent->new,
            hostport     => $hostport,
            guard        => $guard,
        }, $class;
    }

    sub url {
        my ($self) = @_;

        return "https://" . $self->{hostport} . "/";
    }

    sub _request {
        my ($self, $what, $fragment, @args) = @_;

        $self->{lwp}->$what(
            $self->url . $fragment,
            @args,
        );
    }

    sub post   { shift->_request('post',   @_) }
    sub get    { shift->_request('get',    @_) }
    sub put    { shift->_request('put',    @_) }
    sub delete { shift->_request('delete', @_) }
}

sub new_app_and_tester {
    my ($class, $args) = @_;

    local %ENV = %ENV;

    $ENV{$_} = $args->{$_} for keys %$args;

    my $app = do "bin/hopscotch.psgi"
        || die "Failed to execute ./bin/hopscotch: $@\n";

    my $wrapped_app = sub {
        my ($env) = @_;

        my $res = $app->($env);

        Plack::Util::response_cb($res, sub {
            my $res = shift;

            return sub {
                my $chunk = shift;
                return $chunk;
            }
        });
    };

    my $hostport = "$host" . ":" . $port->();

    my $guard = LWP::Protocol::PSGI->register(
        $wrapped_app,
        host => qr/\Q$hostport\E\z/i,
    );

    my $lwp = Hopscotch::Tester::LWP::Wrapper->new(
        $hostport,
        $guard,
    );

    return ($wrapped_app, $lwp);
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "tiny, high-performance HTTP image proxy",
  "AUTHOR" => "Robert Norris",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Hopscotch",
  "EXE_FILES" => [
    "bin/hopscotch.psgi"
  ],
  "LICENSE" => "mit",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "Hopscotch",
  "PREREQ_PM" => {
    "Crypt::Mac::HMAC" => 0,
    "File::LibMagic" => "1.02",
    "Furl::HTTP" => 0,
    "Hash::MultiValue" => 0,
    "List::MoreUtils" => 0,
    "Net::DNS::Paranoid" => 0,
    "Socket" => 0,
    "Try::Tiny" => 0,
    "constant" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "LWP::Protocol::PSGI" => 0,
    "LWP::UserAgent" => 0,
    "MIME::Base64" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Crypt::Mac::HMAC" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::LibMagic" => "1.02",
  "File::Spec" => 0,
  "Furl::HTTP" => 0,
  "Hash::MultiValue" => 0,
  "LWP::Protocol::PSGI" => 0,
  "LWP::UserAgent" => 0,
  "List::MoreUtils" => 0,
  "MIME::Base64" => 0,
  "Net::DNS::Paranoid" => 0,
  "Socket" => 0,
  "Test::More" => "0.96",
  "Try::Tiny" => 0,
  "constant" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Hopscotch/Auth/SHA256.pm
use 5.010;
use strict;
use warnings;
package Hopscotch::Auth::SHA256;

use Crypt::Mac::HMAC qw(hmac_hex);

use constant KEY => $ENV{HOPSCOTCH_KEY} // die "HOPSCOTCH_KEY not set\n";

sub error {
    my ($app, $env, $code, $error) = @_;

    $env->{HOPSCOTCH_AUTH_ERROR} = [ $code, $error ];

    return $app->($env);
}

sub wrap {
    my ($class, $app) = @_;

    return sub {
        my ($env) = @_;

        my (undef, $mac, $hexurl) = split '/', $env->{REQUEST_URI};

        unless (defined $mac && defined $hexurl) {
            return error($app, $env,
                404 => "invalid URL structure (MAC/hex)",
            );
        }

        unless ($hexurl =~ m/^[0-9a-f]+$/) {
            return error($app, $env,
                404 => "invalid characters in hex fragment",
            );
        }

        my $url = pack "h*", $hexurl;
        my $our_mac = hmac_hex('SHA256', KEY, $url);

        unless (lc($mac) eq lc($our_mac)) {
            return error($app, $env,
                404 => "invalid MAC"
            );
        }

        $env->{REQUEST_URI} = $url;

        $app->($env);
    }
}

1;


lib/Hopscotch.pm
use 5.010;
use strict;
use warnings;
package Hopscotch;
# ABSTRACT: tiny, high-performance HTTP image proxy

use Furl::HTTP;
use List::MoreUtils qw(natatime);
use Try::Tiny;
use File::LibMagic 1.02;
use Hash::MultiValue;

use constant HOST          => $ENV{HOPSCOTCH_HOST}          // "unknown";
use constant TIMEOUT       => $ENV{HOPSCOTCH_TIMEOUT}       // 60;
use constant HEADER_VIA    => $ENV{HOPSCOTCH_HEADER_VIA}    // "1.0 hopscotch";
use constant HEADER_UA     => $ENV{HOPSCOTCH_HEADER_UA}     // "hopscotch";
use constant LENGTH_LIMIT  => $ENV{HOPSCOTCH_LENGTH_LIMIT}  // 5242880;
use constant MAX_REDIRECTS => $ENV{HOPSCOTCH_MAX_REDIRECTS} // 4;
use constant PARANOID      => $ENV{HOPSCOTCH_PARANOID}      // 0;
use constant ERRORS        => $ENV{HOPSCOTCH_ERRORS}        // 1;
use constant CAFILE        => $ENV{HOPSCOTCH_CAFILE}        // undef;
use constant CAPATH        => $ENV{HOPSCOTCH_CAPATH}        // undef;
use constant IGNORE_CERTS  => $ENV{HOPSCOTCH_IGNORE_CERTS}  // undef;
use constant NO_WARN       => $ENV{HOPSCOTCH_NO_WARN}       // undef;

my %COPY_REQUEST_HEADERS = map { $_ => 1 } qw(
    accept accept-language cache-control if-modified-since if-match if-none-match if-unmodified-since
);

my %COPY_RESPONSE_HEADERS = map { $_ => 1 } qw(
    content-type cache-control etag expires last-modified content-length server
);

my %VALID_CONTENT_TYPES = map { $_ => 1 } qw(
    image/bmp
    image/bitmap
    image/cgm
    image/g3fax
    image/gif
    image/ief
    image/jp2
    image/jpg
    image/jpeg
    image/pict
    image/png
    image/prs.btif
    image/svg+xml
    image/tiff
    image/webp
    image/vnd.adobe.photoshop
    image/vnd.djvu
    image/vnd.dwg
    image/vnd.dxf
    image/vnd.fastbidsheet
    image/vnd.fpx
    image/vnd.fst
    image/vnd.fujixerox.edmics-mmr
    image/vnd.fujixerox.edmics-rlc
    image/vnd.microsoft.icon
    image/vnd.ms-modi
    image/vnd.net-fpx
    image/vnd.wap.wbmp
    image/vnd.xiff
    image/x-cmu-raster
    image/x-cmx
    image/x-macpaint
    image/x-pcx
    image/x-pict
    image/x-portable-anymap
    image/x-portable-bitmap
    image/x-portable-graymap
    image/x-portable-pixmap
    image/x-quicktime
    image/x-rgb
    image/x-xbitmap
    image/x-xpixmap
    image/x-xwindowdump
);

sub request_headers {
    my ($headers) = @_;
    my $it = natatime 2, @$headers;
    [
        (map { my ($k, $v) = $it->(); exists $COPY_REQUEST_HEADERS{lc($k)} ? ($k, $v) : () } (1..(scalar @$headers)/2)),
        'Via' => HEADER_VIA,
    ];
}

sub response_headers {
    my ($headers) = @_;
    my $it = natatime 2, @$headers;
    [
        (map { my ($k, $v) = $it->(); exists $COPY_RESPONSE_HEADERS{lc($k)} ? ($k, $v) : () } (1..(scalar @$headers)/2)),
        'Via' => HEADER_VIA,
        'X-Hopscotch-Host' => HOST,
    ];
}

sub cleanup_error {
    return unless ERRORS;
    my ($err) = @_;
    $err =~ s/ at .+ line.*//sm;
    "E: $err\n";
}

sub response {
    my $err;
    if (defined $_[2]) {
        $err = cleanup_error($_[2]) if defined $_[2];
        warn $err unless NO_WARN;
        $_[1] //= [];
        push @{$_[1]}, 'Content-type' => 'text/plain';
    }
    [ $_[0], response_headers($_[1] // []), [ $err // () ] ];
}

my %header_checks = (
    "content-length" => sub { 0+$_[0] <= LENGTH_LIMIT },
);

sub rejected {
    my ($code, $msg, $headers) = @_;

    {
        my $codetype = substr $code, 0, 1;
        return (1, "remote returned $code $msg") if $codetype < 2 || $codetype > 3;
    };

    {
        my $it = natatime 2, @$headers;
        while (my ($k, $v) = $it->()) {
            return (1, "remote failed header check '$k' with value '$v'") if exists $header_checks{$k} && !$header_checks{$k}->($v);
        }
    }

    return (0);
}

sub type {
    return [ grep { defined $_ && exists $VALID_CONTENT_TYPES{lc(($_ =~ m/^([^;]+)/)[0] // "")} } @_ ]->[0];
}

my $magic = File::LibMagic->new(uncompress => 1);

sub rewrite_headers_from_body {
    my ($headers, $buf) = @_;

    my $hh = Hash::MultiValue->new(@$headers);

    my $type = type($hh->{'content-type'});
    if (!$type && $buf) {
        $type = type($magic->info_from_string($buf)->{mime_type});
    }
    unless ($type) {
        my $err = sprintf "remote offered non-image content%s\n", $hh->{'content-type'} ? " ($hh->{'content-type'})" : "";
        return ($headers, $err);
    }
    $hh->set('content-type', $type);

    return ([$hh->flatten]);
}

my $furl = Furl::HTTP->new(
    timeout       => TIMEOUT,
    max_redirects => MAX_REDIRECTS,
    agent         => HEADER_UA,
    ssl_opts      => {
        CAFILE       ? (SSL_ca_file => CAFILE)            : (),
        CAPATH       ? (SSL_ca_path => CAPATH)            : (),
        IGNORE_CERTS ? (SSL_verify_callback => sub { 1 }) : (),
    },
);

if (PARANOID) {
    require Net::DNS::Paranoid;
    require Socket;
    my $dns = Net::DNS::Paranoid->new;
    $furl->{inet_aton} = sub {
        my ($host) = @_;
        my ($addrs) = try { $dns->resolve($host) };
        unless (defined $addrs) {
            $! = 14; # EFAULT = Bad address ;)
            return;
        }
        Socket::inet_aton($host);
    }
}

sub to_app {
    return \&app;
}

sub app {
    my ($env) = @_;

    unless ($env->{REQUEST_METHOD} =~ m/^GET|HEAD$/) {
        return response(
            405,
            [],
            "request method must be GET or HEAD (not $env->{REQUEST_METHOD})"
        );
    }

    # Our Auth middleware may set this
    if ($env->{HOPSCOTCH_AUTH_ERROR}) {
        my ($code, $error) = @{ $env->{HOPSCOTCH_AUTH_ERROR} };

        return response($code, [], $error);
    }

    my $url = $env->{REQUEST_URI};

    # Valid chars are unreserved + reserved chars from https://tools.ietf.org/html/rfc3986#section-2.2
    return response(404, [], "invalid characters in URL")
      if $url =~ m{[^A-Za-z0-9\-\._~:\/\?#\[\]\@!\$&'\(\)\*\+,;=\%]}o;

    return sub {
        my ($respond) = @_;

        my $w;
        my $bytes = 0;

        my (undef, $code, $msg, $headers, $body) = try {
            my @return = $furl->request(
                method     => "GET",
                url        => $url,
                headers    => request_headers([map { (substr($_, 5) =~ s/_/-/gr) => $env->{$_} } grep { m/^HTTP_/ } keys %$env]),
                write_code => sub {
                    my ($code, $msg, $headers, $buf) = @_;

                    my ($rejected, $err) = rejected($code, $msg, $headers);
                    die $err if $rejected;

                    return if defined $buf && length $buf == 0;

                    unless (defined $w) {
                        my ($headers_out, $err) = rewrite_headers_from_body($headers, $buf);
                        die "$err\n" if $err;
                        $w = $respond->([@{response($code, $headers_out)}[0,1]]);
                    }

                    if (defined $buf) {
                        if ($env->{REQUEST_METHOD} eq "HEAD") {
                            die []; # magic to force proxy to end successfully
                        }

                        $bytes += length $buf;
                        die "remote file exceeded length limit\n" if $bytes > LENGTH_LIMIT;
                        $w->write($buf);
                    }
                }
            );

            if ($w) {
                # Flush out the response for plack
                $w->close;
            }

            return @return;
        }
        catch {
            if ($w) {
                unless (ref $_) {
                    # part way through stream so we can't return the error to the client
                    warn cleanup_error("stream aborted: $_") unless NO_WARN;
                }
                $w->close;
            }
            else {
                $respond->(response(404, [], $_));
            }
            $w = 1; # flag that we've responded
        };

        unless ($w) {
            my $err;
            return $respond->(response($code, $headers)) if $code == 304;
            (my $rejected, $err) = rejected($code, $msg, $headers);
            if ($rejected) {
                return $respond->(response(404, [], $err));
            }
            (my $headers_out, $err) = rewrite_headers_from_body($headers);
            if ($err) {
                return $respond->(response(404, [], $err));
            }
            return $respond->(response(200, $headers_out));
        }
    };
}

if (my ($url) = @ARGV) {
    $url =~ s{http://[^/]+}{};
    my $r = app({
        REQUEST_METHOD => 'GET',
        REQUEST_URI    => $url,
    });

    if (ref $r eq "ARRAY") {
        # error already output in response()
        exit 0;
    }

    if (ref $r eq "CODE") {
        open my $w, '<', '/dev/null';
        my $respond = sub { $w };
        $r->($respond);
    }

    exit 0;
}

1;
